cmd/util/ledger/migrations/cadence_values_migration.go

```diff
@@ -0,0 +1,291 @@
+
+	m.log.Info().Msg("Migrating cadence values")
+
+	migration.Migrate(
+		&migrations.AddressSliceIterator{
+			Addresses: []common.Address{
+				address,
+			},
+		},
+		migration.NewValueMigrationsPathMigrator(
+			reporter,
+			m.valueMigrations(accounts, reporter)...,
+		),
+	)
+
+	m.log.Info().Msg("Committing changes")
+	err = migration.Commit()
+	if err != nil {
+		return nil, fmt.Errorf("failed to commit changes: %w", err)
+	}
+
+	// finalize the transaction
+	result, err := migrationRuntime.TransactionState.FinalizeMainTransaction()
+	if err != nil {
+		return nil, fmt.Errorf("failed to finalize main transaction: %w", err)
+	}
+
+	// Merge the changes to the original payloads.
+	return m.mergeRegisterChanges(migrationRuntime, result.WriteSet)
+}
+
+func (m *CadenceBaseMigrator) mergeRegisterChanges(
+	mr *migratorRuntime,
+	changes map[flow.RegisterID]flow.RegisterValue,
+) ([]*ledger.Payload, error) {
+
+	originalPayloads := mr.Snapshot.Payloads
+	newPayloads := make([]*ledger.Payload, 0, len(originalPayloads))
+
+	// Add all new payloads.
+	for id, value := range changes {
+		key := convert.RegisterIDToLedgerKey(id)
+		newPayloads = append(newPayloads, ledger.NewPayload(key, value))
+	}
+
+	// Add any old payload that wasn't updated.
+	for id, value := range originalPayloads {
+		if len(value.Value()) == 0 {
+			// This is strange, but we don't want to add empty values. Log it.
+			m.log.Warn().Msgf("empty value for key %s", id)
+			continue
+		}
+
+		// If the payload had changed, then it has been added earlier.
+		// So skip old payload.
+		if _, contains := changes[id]; contains {
+			continue
+		}
+
+		newPayloads = append(newPayloads, value)
+	}
+
+	return newPayloads, nil
+}
+
+func NewCadenceValueMigrator(
+	rwf reporters.ReportWriterFactory,
+	capabilityIDs map[interpreter.AddressPath]interpreter.UInt64Value,
+) *CadenceBaseMigrator {
+	return &CadenceBaseMigrator{
+		reporter: rwf.ReportWriter("cadence-value-migrator"),
+		valueMigrations: func(
+			accounts environment.Accounts,
+			reporter *cadenceValueMigrationReporter,
+		) []migrations.ValueMigration {
+			// All cadence migrations except the `capcons.LinkValueMigration`.
+			return []migrations.ValueMigration{
+				&capcons.CapabilityValueMigration{
+					CapabilityIDs: capabilityIDs,
+					Reporter:      reporter,
+				},
+				string_normalization.NewStringNormalizingMigration(),
+				account_type.NewAccountTypeMigration(),
+				type_value.NewTypeValueMigration(),
+			}
+		},
+	}
+}
+
+func NewCadenceLinkValueMigrator(
+	rwf reporters.ReportWriterFactory,
+	capabilityIDs map[interpreter.AddressPath]interpreter.UInt64Value,
+) *CadenceBaseMigrator {
+	return &CadenceBaseMigrator{
+		reporter: rwf.ReportWriter("cadence-link-value-migrator"),
+		valueMigrations: func(
+			accounts environment.Accounts,
+			reporter *cadenceValueMigrationReporter,
+		) []migrations.ValueMigration {
+			idGenerator := environment.NewAccountLocalIDGenerator(
```

Comment:
I think Janez previously noted that the changes from the account local ID generator must be committed

Reply:
I thought using the `AccountLocalIDGenerator` and committing the transaction state at the end would also commit ID generation changes. @janezpodhostnik, Or do I need to do anything additional here?

Reply:
Ah, I see, sounds good! 👍

I assumed an explicit commit is needed, but I guess there's an implicit commit through committing the state that is used by the accounts/generator

Reply:
> committing the transaction state at the end would also commit ID generation changes

yes that should be the case! But best confirm it wit a test if possible.

Reply:
Added in https://github.com/onflow/flow-go/pull/5192/commits/82365ae865e2142f0b85361ec810da189537751b